FinalLocalVariableCheck: NullPointerException on compact source files (JEP 512)

This is part of the umbrella issue for Java 25 [Compact Source Files and Instance Main Methods](https://github.com/checkstyle/checkstyle/issues/19971)

I have read check documentation: https://checkstyle.org/checks/coding/finallocalvariable.html
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

FinalLocalVariableCheck throws a `NullPointerException` when it processes a JEP 512 compact source file that declares a top-level field. A top-level field is a `VARIABLE_DEF` whose parent is the synthetic `COMPACT_COMPILATION_UNIT`, not `OBJBLOCK`, so the check mistakes it for a local-variable candidate and calls `scopeStack.peek()` while the scope stack is empty.

```
vivek@Viveks-MacBook-Air checkstyle % cat > Example.java <<'EOF'
int field = 1;

void main() {
    System.out.println(field);
}
EOF

vivek@Viveks-MacBook-Air checkstyle % cat > config.xml <<'EOF'
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
    "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
    "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
  <module name="TreeWalker">
    <module name="FinalLocalVariable"/>
  </module>
</module>
EOF

vivek@Viveks-MacBook-Air checkstyle % RUN_LOCALE="-Duser.language=en -Duser.country=US"
java $RUN_LOCALE -jar target/checkstyle-13.5.0-SNAPSHOT-all.jar -c config.xml Example.java
Starting audit...
com.puppycrawl.tools.checkstyle.api.CheckstyleException: Exception was thrown while processing Example.java
        at com.puppycrawl.tools.checkstyle.Checker.processFiles(Checker.java:313)
        at com.puppycrawl.tools.checkstyle.Checker.process(Checker.java:227)
        at com.puppycrawl.tools.checkstyle.Main.runCheckstyle(Main.java:429)
        at com.puppycrawl.tools.checkstyle.Main.runCli(Main.java:347)
        at com.puppycrawl.tools.checkstyle.Main.execute(Main.java:205)
        at com.puppycrawl.tools.checkstyle.Main.main(Main.java:129)
Caused by: java.lang.NullPointerException: Cannot read field "scope" because the return value of "java.util.Deque.peek()" is null
        at com.puppycrawl.tools.checkstyle.checks.coding.FinalLocalVariableCheck.insertVariable(FinalLocalVariableCheck.java:508)
        at com.puppycrawl.tools.checkstyle.checks.coding.FinalLocalVariableCheck.visitToken(FinalLocalVariableCheck.java:219)
        at com.puppycrawl.tools.checkstyle.TreeWalker.notifyVisit(TreeWalker.java:384)
        at com.puppycrawl.tools.checkstyle.TreeWalker.processIter(TreeWalker.java:457)
        at com.puppycrawl.tools.checkstyle.TreeWalker.walk(TreeWalker.java:322)
        at com.puppycrawl.tools.checkstyle.TreeWalker.processFiltered(TreeWalker.java:202)
        at com.puppycrawl.tools.checkstyle.api.AbstractFileSetCheck.process(AbstractFileSetCheck.java:101)
        at com.puppycrawl.tools.checkstyle.Checker.processFile(Checker.java:341)
        at com.puppycrawl.tools.checkstyle.Checker.processFiles(Checker.java:300)
        ... 5 more
Checkstyle ends with 1 errors.
```
---

**Describe what you expect in detail.**

FinalLocalVariableCheck must not crash on compact source files. In a JEP 512 compact source file, top-level fields and methods are wrapped in a synthetic `COMPACT_COMPILATION_UNIT` node (the implicit class), so a top-level field's parent is `COMPACT_COMPILATION_UNIT` rather than `OBJBLOCK`.

The check's `VARIABLE_DEF` handler only skips fields when `parent == OBJBLOCK`; for any other parent it treats the variable as a local candidate and calls `insertVariable`, which dereferences `scopeStack.peek()`. At the top level no method/`OBJBLOCK` scope has been pushed, so `peek()` returns `null` and the `.scope` access throws.

A top-level field in a compact source file is a **field** of the implicit class, not a local variable, so it must be excluded from final-local analysis exactly as `OBJBLOCK`-parented fields are. Expected behaviour:
- no `NullPointerException`;
- no violation reported on the top-level field `field` (it is a field, not a local);
- local variables inside top-level methods are still analysed normally 

---

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FinalLocalVariableCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FinalLocalVariableCheck.java
@@ -142,6 +142,7 @@ public class FinalLocalVariableCheck extends AbstractCheck {
             TokenTypes.METHOD_DEF,
             TokenTypes.SLIST,
             TokenTypes.OBJBLOCK,
+            TokenTypes.COMPACT_COMPILATION_UNIT,
             TokenTypes.LITERAL_BREAK,
             TokenTypes.LITERAL_FOR,
             TokenTypes.EXPR,
@@ -156,6 +157,7 @@ public class FinalLocalVariableCheck extends AbstractCheck {
             TokenTypes.METHOD_DEF,
             TokenTypes.SLIST,
             TokenTypes.OBJBLOCK,
+            TokenTypes.COMPACT_COMPILATION_UNIT,
             TokenTypes.LITERAL_BREAK,
             TokenTypes.LITERAL_FOR,
             TokenTypes.VARIABLE_DEF,
@@ -171,6 +173,7 @@ public class FinalLocalVariableCheck extends AbstractCheck {
             TokenTypes.METHOD_DEF,
             TokenTypes.SLIST,
             TokenTypes.OBJBLOCK,
+            TokenTypes.COMPACT_COMPILATION_UNIT,
             TokenTypes.LITERAL_BREAK,
             TokenTypes.LITERAL_FOR,
             TokenTypes.VARIABLE_DEF,
@@ -184,8 +187,8 @@ public class FinalLocalVariableCheck extends AbstractCheck {
     @Override
     public void visitToken(DetailAST ast) {
         switch (ast.getType()) {
-            case TokenTypes.OBJBLOCK, TokenTypes.METHOD_DEF,
-                 TokenTypes.CTOR_DEF, TokenTypes.LITERAL_FOR ->
+            case TokenTypes.COMPACT_COMPILATION_UNIT, TokenTypes.OBJBLOCK,
+                 TokenTypes.METHOD_DEF, TokenTypes.CTOR_DEF, TokenTypes.LITERAL_FOR ->
                 scopeStack.push(new ScopeData());
 
             case TokenTypes.SLIST -> {

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/FinalLocalVariableCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/FinalLocalVariableCheckTest.java
@@ -462,4 +462,16 @@ public class FinalLocalVariableCheckTest
             getPath("InputFinalLocalVariableInterface.java"),
             expected);
     }
+
+    @Test
+    public void testCompactSourceFile() throws Exception {
+        final String[] expected = {
+            "18:9: " + getCheckMessage(MSG_KEY, "lambdaLocal"),
+            "24:13: " + getCheckMessage(MSG_KEY, "branchLocal"),
+            "31:9: " + getCheckMessage(MSG_KEY, "local"),
+        };
+        verifyWithInlineConfigParser(
+                getNonCompilablePath("InputFinalLocalVariableCompactSourceFile.java"),
+            expected);
+    }
 }

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/finallocalvariable/InputFinalLocalVariableCompactSourceFile.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/finallocalvariable/InputFinalLocalVariableCompactSourceFile.java
@@ -1,0 +1,36 @@
+/*
+FinalLocalVariable
+validateEnhancedForLoopVariable = (default)false
+validateUnnamedVariables = (default)false
+tokens = (default)VARIABLE_DEF
+
+*/
+
+// non-compiled with javac: Compilable with Java25
+
+import java.util.function.Supplier;
+
+int field = 1;
+
+final int constField = 2;
+
+Supplier<Integer> supplier = () -> {
+    int lambdaLocal = 5; // violation, 'Variable 'lambdaLocal' should be declared final'
+    return lambdaLocal;
+};
+
+int categorized = switch (field) {
+    case 1 -> {
+        int branchLocal = 10; // violation, 'Variable 'branchLocal' should be declared final'
+        yield branchLocal;
+    }
+    default -> 0;
+};
+
+void main() {
+    int local = 2; // violation, 'Variable 'local' should be declared final'
+    int reassigned = 0;
+    reassigned = 9;
+    System.out.println(field + constField + local + categorized + reassigned);
+    supplier.get();
+}

